derive_client/clients/base_client.py

```diff
@@ -786,24 +815,25 @@ def get_manager_for_subaccount(self, subaccount_id, asset_name):
         deposit_currency = UnderlyingCurrency[asset_name]
         currency = self.fetch_currency(asset_name)
         underlying_address = currency['protocol_asset_addresses']['spot']
-        manager_addresses = currency['managers']
-
-        if len(manager_addresses) == 1:
-            manager_address = manager_addresses[0].get('address')
-        else:
-            to_account = self.fetch_subaccount(subaccount_id)
-            account_type = (
-                SubaccountType.STANDARD if to_account.get("margin_type") == "SM" else SubaccountType.PORTFOLIO
-            )
-            account_currency = UnderlyingCurrency[to_account.get("currency")]
-            index = (
-                0 if account_type is SubaccountType.STANDARD else 1 if account_currency is UnderlyingCurrency.ETH else 2
-            )
-            manager_address = manager_addresses[index].get('address')
-
-        if not manager_address or not underlying_address:
+        managers = list(map(lambda kwargs: ManagerAddress(**kwargs), currency['managers']))
+        manager_by_type = {}
+        for manager in managers:
+            manager_by_type.setdefault((manager.margin_type, manager.currency), []).append(manager)
+
+        to_account = self.fetch_subaccount(subaccount_id)
+        account_currency = MainnetCurrency[to_account.get("currency")]
+        margin_type = MarginType[to_account.get("margin_type")]
+
+        def get_unique_manager(margin_type, currency):
+            matches = manager_by_type.get((margin_type, currency), [])
+            if len(matches) != 1:
+                raise ValueError(f"Expected exactly one ManagerAddress for {(margin_type, currency)}, found {matches}")
+            return matches[0]
```

Comment:
the order changed; indexing is brittle. Also now there are 5 managers retrieved. We're selecting by margin type and currency, ensuring there is exactly one unique match (as we assume)